sdk/typescript/src/transactions/ObjectCache.ts

```diff
@@ -0,0 +1,325 @@
+	cache?: AsyncCache;
+	address: string;
+}
+
+export class ObjectCache implements TransactionBlockDataResolverPlugin {
+	#cache: AsyncCache;
+	#address: string;
+
+	constructor({ cache = new InMemoryCache(), address }: ObjectCacheOptions) {
+		this.#cache = cache;
+		this.#address = normalizeSuiAddress(address);
+	}
+
+	getObjects: NonNullable<TransactionBlockDataResolverPlugin['getObjects']> = async (ids, next) => {
+		const results = new Map<string, ObjectCacheEntry>();
+
+		const cached = await this.#cache.getObjects(ids);
+
+		cached.forEach((object) => {
+			if (object) {
+				results.set(object.objectId, object);
+			}
+		});
+
+		const missingIds = ids.filter((id) => !results.has(id));
+
+		if (missingIds.length > 0) {
+			const newObjects = await next(missingIds);
+
+			await Promise.all(
+				newObjects.map(async (newObject) => {
+					await this.#cache.addObject(newObject);
+					results.set(newObject.objectId, newObject);
+				}),
+			);
+		}
+
+		return ids.map((id) => results.get(id)!);
+	};
+
+	getMoveFunctionDefinition: NonNullable<
+		TransactionBlockDataResolverPlugin['getMoveFunctionDefinition']
+	> = async (ref, next) => {
+		const cached = await this.#cache.getMoveFunctionDefinition(ref);
+		if (cached) {
+			return cached;
+		}
+
+		const functionDefinition = await next(ref);
+
+		return await this.#cache.addMoveFunctionDefinition(functionDefinition);
+	};
+
+	async clearCache() {
+		await this.#cache.clear();
+	}
+
+	async clearOwnedObjects() {
+		await this.#cache.clear('OwnedObject');
+	}
+
+	async applyEffects(effects: typeof bcs.TransactionEffects.$inferType) {
+		if (!effects.V2) {
+			throw new Error(`Unsupported transaction effects version ${effects.$kind}`);
+		}
+
+		const { lamportVersion, changedObjects } = effects.V2;
+
+		await Promise.all(
+			changedObjects.map(async ([id, change]) => {
+				if (change.outputState.NotExist) {
+					await this.#cache.deleteObject(id);
+				} else if (change.outputState.ObjectWrite) {
+					const [digest, owner] = change.outputState.ObjectWrite;
+
+					// Remove objects not owned by address after transaction
+					if (owner.ObjectOwner || (owner.AddressOwner && owner.AddressOwner !== this.#address)) {
+						await this.#cache.deleteObject(id);
+					}
+
+					await this.#cache.addObject({
+						objectId: id,
+						digest,
+						version: lamportVersion,
+						owner: owner.AddressOwner ?? owner.ObjectOwner ?? null,
+						initialSharedVersion: owner.Shared?.initialSharedVersion ?? null,
+					});
+				}
+			}),
+		);
+	}
+}
+
+export class CachingTransactionBlockExecutor {
+	#client: SuiClient;
+	#protocolConfig: ProtocolConfig | null = null;
+	cache: ObjectCache;
+
+	constructor({
+		client,
+		...options
+	}: ObjectCacheOptions & {
+		client: SuiClient;
+	}) {
+		this.#client = client;
+		this.cache = new ObjectCache(options);
+	}
+
+	async #getProtocolConfig() {
+		if (!this.#protocolConfig) {
+			this.#protocolConfig = await this.#client.getProtocolConfig();
```

Comment:
Technically this should only be cached for one epoch but idk the best way to represent that. This is probably fine.